[quant][graphmode][fx] Add reference option support for binary ops

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63828
* #63627
* #63501
* #62861
* __->__ #62698

Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

jerryzh168 committed:
[quant][graphmode][fx] Add reference option support for binary ops

Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

vkuzo commented on torch/quantization/fx/convert.py:
```diff
@@ -335,11 +335,11 @@ def node_arg_is_quantized(node_arg: Any) -> bool:
             return False
 
-    def is_output_quantized(node: Node, obj: QuantizeHandler, qconfig: QConfigAny, modules: Dict[str, torch.nn.Module]) -> bool:
+    def is_output_quantized(node: Node, obj: QuantizeHandler, qconfig: QConfigAny, modules: Dict[str, torch.nn.Module], is_reference=False) -> bool:
```
could we give an example why `is_reference` should matter when deciding whether the output is quantized?

jerryzh168 replied:
sure, this is because when is_reference is False, we would return float node when a certain dtype combination is not supported (since fbgemm/qnnpack only support certain dtype combinations), so the output may be float, but when is_reference is True, we support all dtype combinations so the output will always be quantized

vkuzo replied:
makes sense.  Is this the behavior we want?  Should we have a different flag to control this?  It seems unexpected that setting is_reference to True can change what is being quantized.

jerryzh168 replied:
hmm, this is an intermediate state, I think the alternative is that in case when the dtype combination is not supported when is_reference=False, we will produce the reference pattern. I think that is more aligned with what we'll have in the future.

vkuzo replied:
I think it's worth considering having `is_reference` only mean one thing (quantized ops are in reference format), and introduce a new setting to control what happens to ops whose quantization is not defined for fbgemm/qnnpack.  Otherwise, `is_reference` starts to have multiple meanings, it gets harder to explain, it starts to potentially change numerical outputs, and it grows in scope.

What are your thoughts?

jerryzh168 replied:
Yeah I understand that `is_reference` in the current state is not a very well defined flag. It's not the final state though, in terms of removing the ambiguity I think it might be easier just to remove this flag, we can do that gradually for each quantizehandler and then in the end remove it. I'm thinking of doing that, it's just not very high priority for me currently. Would we have someone that can help with this?

jerryzh168 replied:
so in the future, the semantics will be, we'll produce the pattern by default, if it's not supported by fbgemm, it will stay as a reference pattern and not be fused. does that sound good?

vkuzo replied:
>  It's not the final state though, to remove the ambiguity I think it might be easier just to remove this flag, we can do that gradually for each quantizehandler(one example is shown in next PR) and then in the end remove it.

makes sense if it's not the most urgent, adding a TODO sounds fine.  However, it seems like a p0 to me to get this done before rolling is_reference out to real customers.

jerryzh168 replied:
yeah, I think we should do it before the beta release

vkuzo replied:
> yeah, I think we should do it before the beta release

IMO this should be done before any internal customers use the code, so before prototype release :). We should discuss and come up with something which works for everyone.

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -1239,6 +1261,13 @@ def __init__(
         elif node.op == "call_module":
             self.op = type(modules[str(node.target)])
 
+    def is_output_quantized(self, qconfig, is_reference):
```
This looks like a dup of the same function earlier

jerryzh168 replied:
yes, and binary_op_supported_dtypes is replaced with default_op_supported_dtypes

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

raghuramank100 commented on test/quantization/fx/test_numeric_suite_fx.py:
```diff
@@ -662,6 +661,9 @@ def _op_is_unmatchable(op):
                 pass
             elif qhandler_cls == qp.DefaultNodeQuantizeHandler:
+                # torch.sum does not have quantized equivalents
+                if base_op == torch.sum:
```
For context, did we support torch.sum previously and is that being removed now?

jerryzh168 replied:
no, torch.sum is skipped previously as well, we changed the quantize handler from BinaryOpQuantizeHandler to DefaultNodeQuantizeHandler so the skip code is moved here

vkuzo replied:
I can see this getting pretty hard to understand.  With this rule, there seem to be at least three sets of ops:
1. quantization supported by fbgemm/qnnpack, and reference ops
2. quantization supported by reference ops
3. quantization not supported at all

Is there any place where users can find which ops belong to which set?  Would it be possible to move this to a mapping, so the information is easy to understand?  Something like

```
op_to_supported_backends = {
  torch.sum: None,
  torch.add: {BACKEND_FBGEMM, BACKEND_QNNPACK, BACKEND_REFERENCE},
  ...
}
```

vkuzo replied:
above fine to be done in a future PR

jerryzh168 replied:
yeah that's true, we don't have this right now, will make this clearer to understand in the future. Basically quantization will be configurable and the configuration will be specific to backend, so we'll not be tracking what is supported by reference ops in the future and we'll only need to track ops supported by some specific backend.

raghuramank100 commented on torch/quantization/fx/convert.py:
```diff
@@ -335,11 +335,13 @@ def node_arg_is_quantized(node_arg: Any) -> bool:
         assert modules is not None
         # by default the output for a quantizable node is expected to be quantized
-        quantized = True
+        quantized = obj.is_output_quantized(qconfig, is_reference)
```
Is the comment now out of date? It looks like we check this for each quantize handler object.

jerryzh168 replied:
Oh OK, yeah will update

raghuramank100 commented on torch/quantization/fx/match_utils.py:
```diff
@@ -135,60 +131,15 @@ def record_match(pattern, node, matched):
+                            qconfig_map[n.name])
+                        all_matched.add(n.name)
+                    # break after finding the first match
+                    break
```
Can you provide more conext on how you were able to remove all the code here?

jerryzh168 replied:
since reference op supports all dtype combinations, `supported_is_reference` is always True and we always need to match the pattern, in case is_reference=False and the dtype combination is not supported by fbgemm/qnnpack we would still insert observers but will produce a floating point Tensor currently

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30093781).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -180,34 +189,52 @@ def convert(self,
-    torch.div: binary_op_float16_dtypes,
-    operator.truediv: binary_op_float16_dtypes,
-    torch.sum: binary_op_float16_dtypes
+    operator.add: all_dtypes,
+    torch.add: all_dtypes,
+    operator.mul: all_dtypes,
+    torch.mul: all_dtypes,
+    torch.bmm: fp16_dtypes,
+    torch.sub: fp16_dtypes,
+    operator.sub: fp16_dtypes,
+    torch.div: fp16_dtypes,
+    operator.truediv: fp16_dtypes,
 }
```
should this variable now be named something like `fbgemm_and_qnnpack_binary_op_supported_dtypes`?

jerryzh168 replied:
this one actually contains ops that's not supported by fbgemm/qnnpack as well (fp16 ops) I guess this does not make much sense, we should probably be removing fp16_dtypes mapping here..

jerryzh168 replied:
added a comment for these, we'll remove them from this file some time in the future

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -180,34 +189,52 @@ def convert(self,
-    operator.mul: binary_op_int8_dtypes,
-    torch.mul: binary_op_int8_dtypes,
+
+default_op_supported_dtypes = {
```
and this one

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add reference option support for binary ops"


Summary:
We also removed the special handling in match_utils for binary ops

Test Plan:
python test/test_quantize.py TestQuantizeFx
python test/test_quantize.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30093781](https://our.internmc.facebook.com/intern/diff/D30093781)

[ghstack-poisoned]
